File: nvpl_blas/c/cher2.c
/******************************************************************************
 * Content: 
 *     This example demonstrates use of API as below:
 *     cblas_cher2
 *
 ******************************************************************************/
#include "example_helper.h"

int main() {
    nvpl_int_t N = 2;
    nvpl_int_t lda = N;
    nvpl_int_t incX = 1;
    nvpl_int_t incY = 1;
    nvpl_scomplex_t alpha = {1.0f, 2.0f};
    enum CBLAS_ORDER order = CblasRowMajor;
    enum CBLAS_UPLO uplo = CblasUpper;
    nvpl_scomplex_t * A = NULL;
    nvpl_scomplex_t * X = NULL;
    nvpl_scomplex_t * Y = NULL;
    nvpl_int_t len_x = 1 + (N - 1) * labs(incX);
    nvpl_int_t len_y = 1 + (N - 1) * labs(incY);

    printf("\nExample: cblas_cher2 for the Hermitian rank-2 update\n\n");
    printf("#### args: n=%" PRId64 ", lda=%" PRId64 ", incX=%" PRId64 ", incY=%" PRId64 ", order=%c, uplo=%c\n",
           (int64_t)N, (int64_t)lda, (int64_t)incX, (int64_t)incY, order_to_char(order), uplo_to_char(uplo));
    printf("           alpha=(%g, %g)\n", alpha.real, alpha.imag);

    // allocate memory
    A = (nvpl_scomplex_t *)malloc(lda * N  * sizeof(nvpl_scomplex_t));
    X = (nvpl_scomplex_t *)malloc(len_x * sizeof(nvpl_scomplex_t));
    Y = (nvpl_scomplex_t *)malloc(len_y * sizeof(nvpl_scomplex_t));

    // fill data
    fill_cmatrix(A, N, N, lda, order, UPLO_TO_FILL_MODE(uplo), CblasNonUnit);
    fill_cvector(X, N, incX);
    fill_cvector(Y, N, incY);

    // print input data
    print_cmatrix(A, N, N, lda, "A", order);
    print_cvector(X, N, incX, "X");
    print_cvector(Y, N, incY, "Y");

    // call cblas_cher2
    cblas_cher2(order, uplo, N, &alpha, X, incX, Y, incY, A, lda);

    // print result
    printf("\nAfter cher2 operation:\n");
    print_cmatrix(A, N, N, lda, "A", order);

    // release memory
    free(A);
    free(X);
    free(Y);
    return 0;
}

File: nvpl_blas/f77/cher2.f
!******************************************************************************
! Content:
!     This example demonstrates use of API as below:
!     CHER2 
!
!******************************************************************************
      program   CHER2_MAIN
*
      implicit none

      include "nvpl_blas_f77_blas.fi"

*       Intrinsic Functions
      intrinsic  abs

      integer         n, incx, incy, i, lda
      parameter       (n=2, incx=1, incy=1, lda=2)
      complex         alpha
      parameter       (alpha=(1.0, 2.0))
      integer         xsize, ysize
      parameter       (xsize=1+(n-1)*abs(incx),
     &                 ysize=1+(n-1)*abs(incy))
      complex         x(xsize), y(ysize), a(lda,n)

*       External Subroutines
      external        print_cvector, fill_cvector, fill_cmatrix,
     &                print_cmatrix
      character*1     uplo
      parameter       (uplo='U')
*
*      Executable Statements
*
      print*
      print 99
*
*       Print input data
      print*
      print 100, n, lda, incx, incy
      print 101, alpha
      call fill_cvector(x,n,incx)
      call fill_cvector(y,n,incy)
      call fill_cmatrix(a,n,n,lda,uplo,'N')
      call print_cmatrix(a,n,n,lda,'A')
      call print_cvector(x,n,incx,'X')
      call print_cvector(y,n,incy,'Y')
*
*      Call CHER2 subroutine
      call CHER2(uplo,n,alpha,x,incx,y,incy,a,lda)
*
      print*
      print 102
      call print_cmatrix(a,n,n,lda,'A')

      stop
 99   format('Example: CHER2 for the Hermitian rank-2 update')
 100  format('#### args: n=',i1,', lda=',i1,', incx=',i1,
     &       ', incy=',i1)
 101  format(11x,'alpha=',f3.1)
 102  format('The result of CHER2: ')
      end
